### DISCUSSION TITLE
OAuth post login redirects

### DISCUSSION THREAD

**Discussion by gregbty** [sergiodxa/remix-auth#106] [Ideas]
Usually when a user navigates to a route that requires authentication, the application will redirect to some login page with a redirection parameter that corresponds to the original url the user attempted to access. For local based authentication, it's trivial to pluck this parameter from the URL (or append to a hidden input field) and properly redirect the user to the appropriate page post login.

However, for OAuth/OpenID Connect based authentication flows, this is a bit more complicated. Providers like Auth0 recommend storing this information in `state` (https://auth0.com/docs/configure/attack-protection/state-parameters#redirect-users) and is what I've typically done in the past with other frameworks. Unfortunately, it seems that `remix-auth` only provides a static `successRedirect` parameter which the developer must specify upfront.

I’ve thought through a potential approach for solving this with `remix-auth`:

`successRedirect` and `failureRedirect` are optional parameters on `authenticate`. If these values are passed in during the login process, the library could serialize the parameters into the `state` parameter that is sent to the OAuth provider:

```javascript
  // login.tsx
  
  authenticator.authenticate('auth0', request, {
    successRedirect: url.searchParams.get('redirectTo') || '/',
    failureRedirect: '/'
  })
```

After verifying the `state`, the library could then automatically redirect to the appropriate paths based on those serialized values (or ignore them if they are explicitly passed into the secondary authenticate call in the callback route). 

Thoughts?

**Comment by sergiodxa** [comment]
I recommend you to store the redirectTo in the session before calling `authenticator.authenticate` and then in the callback route read from there.

**Reply by gregbty** [reply]
That makes sense. I'll try using that approach instead.

**Reply by ngbrown** [reply]
@gregbty Did you get an example working with the suggested approach?

 I wanted to avoid using storage backed sessions before the user had completed a login. And there's a race condition if multiple tabs timeout and revert to the login, the last one into the session will win. So I ended up using a cookie that is set as late as possible.

I also had to work around how the `authenticator.authenticate` throws so I could set the cookie. Unless this functionality is integrated into `remix-auth`, the same sort of work around would be needed for setting the header cookie with sessions.

The page that needs authenticated:

```ts
export const loader: LoaderFunction = async ({ request }) => {
  const urlParams = new URLSearchParams([
    ["deep_link", urlRelativeToRoot(request.url)],
  ]);
  const user = await authenticator.isAuthenticated(request, {
    failureRedirect: `/login?${urlParams.toString()}`,
  });
}

function urlRelativeToRoot(url: string): string {
  return url.replace(/^(?:\/\/|[^/]+)*/, "");
}
```

Then the login page:

```tsx
export default function Login() {
  const [searchParams] = useSearchParams();
  const deep_link = searchParams.get("deep_link");

  return (
    <Form action="/auth/microsoft" method="post">
      <button>Login with Microsoft</button>
      <input
        hidden={true}
        readOnly={true}
        name="_deep_link"
        value={deep_link ?? ""}
      />
    </Form>
  );
}
```

Then the `cookies.server.ts`:

```ts
export const deepLinkCookie = createCookie(`_deep_link_after_login`, {
  maxAge: 5 * 60,
  httpOnly: true,
  secure: true,
  sameSite: "none",
});
```

Then the auth page:

```ts
export let action: ActionFunction = async ({ request }) => {
  const formData = await request.formData();
  const deep_link = formData.get("_deep_link");

  try {
    await authenticator.authenticate("msal", request);
  } catch (e) {
    if (isResponse(e) && isRedirectResponse(e)) {
      const setDeepLinkCookie = await deepLinkCookie.serialize(deep_link, {
        path: "/auth/microsoft/callback",
      });
      e.headers.append("Set-Cookie", setDeepLinkCookie);
    }

    throw e;
  }
};
```

And finally the callback action has to read back the cookie and pass the path to the `authenticate` function:

```ts
export let action: ActionFunction = async ({ request }) => {
  let successRedirect = await deepLinkCookie.parse(
    request.headers.get("Cookie")
  );
  if (typeof successRedirect !== "string" || successRedirect[0] !== "/") {
    successRedirect = "/";
  }
  const urlParams = new URLSearchParams([["deep_link", successRedirect]]);

  return authenticator.authenticate("msal", request, {
    failureRedirect: `/login?${urlParams.toString()}`,
    successRedirect,
  });
};
```

Each callback will need similar code so it might be better to have this functionality as part of the core library. Or there may be better ways to hook into the success, failure redirects and manipulate the redirect and session before it is committed.

**Reply by sergiodxa** [reply]
@ngbrown I wrote about how I added a returnTo feature to a Remix app using Remix Auth https://sergiodxa.com/articles/add-returnto-behavior-to-remix-auth

**Comment by kwiat1990** [comment]
Hi @sergiodxa, I have implemented the redirect based on your code (https://sergiodxa.com/articles/add-returnto-behavior-to-remix-auth). I have use the last code block (custom login page with some button for providers) and it does work fine for login. I have noticed an issue while logging out tough.

For this I have a simple form which calls this action:

```ts
export let action: ActionFunction = async ({ request }) => {
  const url = new URL(request.url);
  const redirectTo = url.searchParams.get("redirectURL") ?? "/";
  await authenticator.logout(request, { redirectTo });
};
```

It does work as expected for e.g. Google but not for Facebook. The latter appends `#_=_` to the callback URL, so I get something like `https://lorem.com/articles/some-article#_=_`. Here starts the weird part: on click auf logout button the action is called, `_session` cookie gets deleted and that's it. Frontend doesn't get updated data, so I need either to reload the page or click on logout button one more time. Then the frontend will be updated (loader from `root.tsx` will be called).

It seems like the hash at the end of the URL breaks things.

**Reply by sergiodxa** [reply]
The `authenticator.logout` function just destroy the session and redirect somewhere, it doesn't ad anything else to the URL, so I'm not sure why you got `#_=_` appended to the URL. Try to open an issue on the Facebook strategy repository.

**Comment by kwiat1990** [comment]
@sergiodxa it‘s a Facebook thing and has nothing to do with a strategy provider or an auth plugin. It looks the same with Next Auth. 

I get indeed redirected but loaders are not called on the first redirect (fron the URL with hash to the same with out it). 

I‘ve played with it a little bit and it generally a problem with a hash in the URL. It can be a Remix issue.

**Comment by odili** [comment]
@sergiodxa I am trying to adapt your current solution for `returnTo` to be used in a multi-strategy implementation (here, Google and Form) but I seem to be hitting the walls. You propose calling the `login` function in the `loader/action` of the login page as well at the `callBack` URL of the OAuth2 strategy.

However the `loader` of the login page seems to be where my confusion lies.  
```ts
import { LoaderFunction, json } from "remix";
import { auth } from "~/services/auth.server";
import { returnToCookie } from "~/services/cookies.server";

export let loader: LoaderFunction = async ({ request }) => {
  let url = new URL(request.url);
  let returnTo = url.searchParams.get("returnTo");

  let headers = new Headers();
  if (returnTo) {
    headers.append("Set-Cookie", await returnToCookie.serialize(returnTo));
  }

  let data = await getData(request);

  return json(data, { headers });
};
```

For `form` strategy I need to call `auth.isAuthenticated` in the login route while for `OAuth2` strategy I need call `auth.authenticate`(login function). The request data coming into login page with `GET` method does not have any `formdata` to indicate strategy to use like I implemented in my login action.
From your code above I assume  `let data = await getData(request);` is getting some `formdata`, but I don't **understand** how.

Here is my current `action` for login route:

```ts
export const action: ActionFunction = async ({request}) => {
  const formData = await request.clone().formData();
  const strategy = formData.get('strategy') as string;

  return await strategyLogin(strategy, request);
};
```

Here is my current `loader` for login route:
```ts
export const loader: LoaderFunction = async ({request}) => {
  const returnTo = new URL(request.url).searchParams.get('returnTo');

  const headers = new Headers();
  if (returnTo) {
    headers.append('Set-Cookie', await returnToCookie.serialize(returnTo));
  }
  const user = await auth.isAuthenticated(request);

  if (user) return redirect(`/u/${user.name}`);

  const session = await sessionStore.getSession(request.headers.get('Cookie'));
 
  const error = session.get(auth.sessionErrorKey);
  return error
    ? json<LoaderData>(
        {error: JSON.parse(error.message)},
        {headers: returnTo ? headers : undefined}
      )
    : {};
};
```
This loader works for `form` strategy but not for any OAuth2 strategy.
I appreciate any guidance. 

Thank you.
